### PR TITLE
Psalm integration

### LINKED ISSUE #47: Switch to vimeo/psalm
On our [last TSC meeting](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling) we agreed to use `vimeo/psalm` as a static analysis tool.

We'd like to switch to use this library instead of phpstan.

Example implementation and configuration can be found in:
- https://github.com/laminas/automatic-releases
- https://github.com/laminas/laminas-cli

### LINKED ISSUE #49: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `psalm.xml` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require --dev vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by batumibiz**
Resolves #47, resolves #49

Basically I did everything as required, but with some changes that I found useful:
- Since the analysis of Psalm displayed too many errors (more than 1k), it would be wiser to make their correction into a separate task. Therefore, in the Psalm config file, I created a temporary block with a list of ignored files that had problems.
Subsequently, problematic files can be gradually removed from the ignore and errors in them can be corrected.
- I did not add the launch of Psalm to Travis-CI, but instead created a separate GitHub Actions workflow for it.
This provides more detailed reports and allows you to separate static analysis from unit testing.
In addition, a separate badge has been added showing the status of analytics.
- I also separated the launch of PHP CodeSniffer from the unit tests and placed it together with the static analyzer.

I have not fixed any source code, this is still a separate task.

### CODE REVIEW

**Comment by batumibiz** [comment]
Additionally, we may not use Travis-CI at all and transfer unit testing to GitHub Action as a separate workflow.

**Comment by Ocramius** [comment]
> Additionally, we may not use Travis-CI at all and transfer unit testing to GitHub Action as a separate workflow.

That's a good idea, but I would suggest skipping that improvement for now, as we're already amidst this big migration :)

File: psalm.xml.dist [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        xmlns="https://getpsalm.org/schema/config"
+        xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
+>
+    <projectFiles>
+        <directory name="bin"/>
+        <directory name="src"/>
+        <directory name="test"/>
+        <ignoreFiles>
+            <directory name="vendor"/>
+            <!-- Temporary suppression of error output -->
```
Comment (Ocramius) [inline psalm.xml.dist]: Hmm, these should be in a psalm baseline - not suppressed.

See https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#using-a-baseline-file

The idea is that a baseline prevents introducing even more issues.

**Review by Ocramius — DISMISSED** [review]
Requires revert to travis-ci (for now) and baseline file.

Otherwise, leave it as-is and we wait for TSC discussion about using github actions for all repos (see https://github.com/laminas/technical-steering-committee/commit/027773e67aae54894f33dd47c12e53521cb5773d - coming up in a couple weeks)

File: psalm.xml.dist (hunk 1 above)
Comment (batumibiz) [reply psalm.xml.dist]: > Hmm, these should be in a psalm baseline - not suppressed.
> See https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#using-a-baseline-file
> The idea is that a baseline prevents introducing even more issues.

Thanks a lot for the link, I missed something at this point. I'll fix it now.

**Comment by batumibiz** [comment]
Moved Psalm Claims List to your-baseline.xml
However on this I have a question ...

- The list of temporarily ignored files with errors - 53 lines (files) in total. The list is simple, easy to track down the author and reconcile corrections with him.
- At the same time, the file your-baseline.xml contains 1833 lines, at a cursory glance at which you want to get drunk and not come to the code for a month :)

It seems to me that it will be easier to fix with a list of files (53)

**Comment by Ocramius** [comment]
> It seems to me that it will be easier to fix with a list of files (53)

The baseline is not managed manually, but through automation (`vendor/bin/psalm --update-baseline`).

If you suppress issues per-file, what happens is that regressions are not picked up :-)

**Comment by batumibiz** [comment]
> > It seems to me that it will be easier to fix with a list of files (53)
> 
> The baseline is not managed manually, but through automation (`vendor/bin/psalm --update-baseline`).
> 
> If you suppress issues per-file, what happens is that regressions are not picked up :-)

I meant that for each file, if necessary, you can create a task. This is much easier than figuring out the 1833 strings.
Yes, they are created automatically, but you will have to fix it manually.

**Comment by Ocramius** [comment]
> Yes, they are created automatically, but you will have to fix it manually.

that will have to be done anyway - the baseline tells us how many issues we fix at each iteration (suppression doesn't).

BTW, I use this same approach on projects with 100000+ errors, and it really brings value when you see if a patch improves or worsens the situation :D

**Comment by boesing** [comment]
> BTW, I use this same approach on projects with 100000+ errors, and it really brings value when you see if a patch improves or worsens the situation :D

So how do you get psalm to notify you that there have been fixes for already ignored errors? afaik, it does not complain about ignored messages as phpstan does when setting [`reportUnmatchedIgnoredErrors`](https://phpstan.org/user-guide/ignoring-errors#reporting-unused-ignores)

**Comment by batumibiz** [comment]
It all depends on the organization of work.
The list of 53 files will be easy to distribute among the performers who will deal with the fixes.

At the same time, you will spend a LOT of time going through the list of 1833 lines and then deciding who and what to assign.

However, here you know better.
For my part, I will gladly try to help as much as I can.

**Comment by batumibiz** [comment]
Revert Travis-CI, or leave it as it is and wait for the end of the discussion of the transition to GitHub Actions?
I did not touch the unit testing, it remained on Travis-CI as it was. Only static analysis is hosted on GitHub Actions.

**Comment by Ocramius** [comment]
> So how do you get psalm to notify you that there have been fixes for already ignored errors? a

in the CI run, it says how many issues have been fixed, if some are no longer in the baseline.

 > Revert Travis-CI, or leave it as it is and wait for the end of the discussion of the transition to GitHub Actions?

IMO leave as-is for now, and we hold the patch until the discussion has happened

**Comment by weierophinney** [comment]
I've rebased your branch off current 3.4.x, which has the GHA workflow. I removed the commits where you added your "analyse" workflow, and I've applied fixes as flagged by Psalm (as there were other changes in the branch that now resulted in errors).

**Pull request merged**